share/smack/include/pthread.h

```diff
@@ -0,0 +1,71 @@
+//
```

Comment:
Mike, to fill you in, Zvonimir and I talked a bit about this, and we didn't really come up with a clear winner...  

The pthread specification only specifies an API and behavior for the API functions.  The internal workings and data structures are left up to the individual implementation.  This means, for example, that struct fields on `pthread_t` objects can be named differently across different implementations.  As a result, our implementation in `pthread.c` can't assume a particular layout or field names for the pthread structs needed to track the status of things.

Further, the system provided header files are more complex, both in terms of struct sizes/complexity and in terms of nested headers/dependencies.  Some of these nested includes can bring in dozens of other header files that wouldn't be used by our implementation.  I don't see this item as being inherently problematic, but rather just unnecessary (and possibly performance degrading).

With that said, our workaround does have some issues. We are preempting the system's `pthread.h` file by passing the path to our version of `pthread.h` on the clang command line.  However, `pthreadtypes.h` is not consistently named across various implementations.  (Sometimes the statement is `#include <pthreadtypes.h>, while other systems may use `#include <bits/pthreadtypes.h>`, while still others may use different folder structures/naming conventions).  As a result, we either need to have a copy of our pthread typedefs in the include paths expected by each implementation, or we need to "preempt" the inclusion of the systems pthread typedefs.

I used include guards to prevent the pthread structs from being defined twice (see https://github.com/smackers/smack/blob/94617b24f5c8579ae8b58e5969bc4e7db9f7d5ae/share/smack/include/pthreadtypes.h#L2 and https://github.com/smackers/smack/blob/94617b24f5c8579ae8b58e5969bc4e7db9f7d5ae/share/smack/include/pthreadtypes.h#L3).

These include guards work, but only if `pthread.h` is included in the user's program *before* any other header that recursively includes the pthread typedefs.  If header `foo.h` is included *before* `pthread.h` in the user's program, and `foo.h` recursively includes `pthreadtypes.h` somehow, then the data structures defined in the system's version of `pthreadtypes.h` will be used, which breaks our implementation in `pthread.c`.

Moving to your solution would do away with this problem, however, we would have to address the issue of unpredictable struct fields (and possibly somehow mitigate the inclusion of unneeded header files).

I am definitely open to any suggestions to address this problem - we just tucked it under the rug so I could make progress on the implementation itself.

I hope this has made sense.  It gets a little bit cumbersome to describe this problem.